File: src/main/java/ADemo.java
/**
 * @Author lzc
 * @Date 2019-05-22 10:51
 */
public class ADemo {
    public static void main(String[] args) {
        A a = new B();
        B b = (B) a;
        System.out.println(a.i);  // 20  X
        System.out.println(b.i);  // 20

        System.out.println(a.getI());
        System.out.println(b.getI());


    }
}

class A{
    static {

    }

    {
      //先与构造函数执行
    }
    public int i = 10;

    public int getI(){
        return i;
    }
}

class B extends A{

    public int i = 20;
    public int getI(){
        return i;
    }
}

File: src/main/scala/com/atguigu/day03/imp/ADemo.scala
package com.atguigu.day03.imp

/**
  * Author lzc
  * Date 2019-05-24 15:32
  */
object B {
}

object A {
    implicit def b2a(b: B) = new A()
    
}

class A {
    
    def afoo() = {
        println("afoo")
    }
}

class B {
    
    def bfoo() = {
        println("bfoo")
        
    }
}

object ADemo {
    
    
    def main(args: Array[String]): Unit = {
        // 接收b类型的对象, 返回一个对象, 这个对象应用有一个方法 afoo()
//        new B().afoo()
        
        val a: A = new B()
        a.afoo()
    
        val set = Set(10, 20, 30)
    }
}



